### ISSUE TITLE
[Serve] Bug in Serve node memory-related resources calculation

### ISSUE THREAD

**Issue by muhammadsyahyas** [ray-project/ray#11198]
### What is the problem?

Initializing ray `ray.init()` with `_memory` or `object_memory_store` kwargs will produce confusing `ray.nodes()` "memory" and "object_store_memory" values information. Initializing without that kwargs will also produce confusing values. As [documented](https://docs.ray.io/en/latest/package-ref.html#ray-init), `_memory` or `object_memory_store` kwargs of `ray.init()` are in **bytes**, however printed node information shows wrong and very different value, even I tried to convert that value into kilobytes, megabytes, or gigabytes.
That leads to another error when I tried to create a backend in Serve and use `memory` or `object_memory_store` in `ray_actor_options` in `client.create_backend()`. It seems that Serve presumed my node did not have enough memory or object store memory to start the actor. However, using non-memory related kwargs, e.g. `num_cpus`, was fine if not exceeding system limitation.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Python 3.7.9, Ray 1.0.0, Ubuntu 18.04 LTS, a laptop with 8 GB RAM

### Reproduction (REQUIRED)
Please provide a script that can be run to reproduce the issue. The script should have **no external library dependencies** (i.e., use fake or mock data / environments):
```python
import ray
from ray import serve
import requests

ray.init(_memory=1024*1024*1024, object_store_memory=1024*1024*1024)
print(ray.nodes())
# > [{ … removed for clarity ..., 'Resources': {'memory': 20.0, 'node:192.168.43.166': 1.0, 'accelerator_type:GT': 1.0, 'object_store_memory': 14.0, 'CPU': 8.0, 'GPU': 1.0}}]
# without _memory and object_store_memory > [{..., 'Resources': {'object_store_memory': 19.0, 'GPU': 1.0, 'accelerator_type:GT': 1.0, 'node:192.168.43.166': 1.0, 'memory': 56.0, 'CPU': 8.0}}] 

client = serve.start()

def echo(flask_request):
    return "hello " + flask_request.args.get("name", "serve!")

client.create_backend(
    "hello", echo, ray_actor_options={"object_store_memory": 128*1024*1024})
# > Traceback A (see below this code block)

client.create_endpoint("hello", backend="hello", route="/hello")

print(client.list_backends())
print(requests.get("http://127.0.0.1:8000/hello").text)
```
Traceback A:
```
Traceback (most recent call last):
  File "ray_serve_error.py", line 15, in <module>
    "hello", echo, ray_actor_options={"object_store_memory": 128*1024*1024})
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/api.py", line 29, in check
    return f(self, *args, **kwargs)
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/api.py", line 246, in create_backend
    replica_config))
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/worker.py", line 1428, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(RayServeException): ray::ServeController.create_backend() (pid=17726, ip=192.168.43.166)
  File "python/ray/_raylet.pyx", line 484, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 435, in ray._raylet.execute_task.function_executor
  File "python/ray/_raylet.pyx", line 1404, in ray._raylet.CoreWorker.run_async_func_in_event_loop
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 428, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 784, in create_backend
    raise e
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 781, in create_backend
    self._scale_replicas(backend_tag, backend_config.num_replicas)
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 554, in _scale_replicas
    num_possible, current_num_replicas + num_possible))
ray.serve.exceptions.RayServeException: Cannot scale backend hello to 1 replicas. Ray Serve tried to add 1 replicas but the resources only allows 0 to be added. To fix this, consider scaling to replica to 0 or add more resources to the cluster. You can check avaiable resources with ray.nodes().
```

If we cannot run your script, we cannot fix your issue.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/latest/installation.html).

**Comment by rkooo567** [comment]
https://github.com/ray-project/ray/issues/6021#issuecomment-546563749

It is because our memory is represented as a 50MB chunk internally. He mentioned it will be fixed in the future, and that future is actually coming soon (we are developing a new distributed scheduler now, and that once that project is done, this will be fixed. It will be likely done within 3 months).

**Comment by rkooo567** [comment]
cc @edoakes

**Comment by muhammadsyahyas** [comment]
@rkooo567 Thank you for the clarification. I hope that will be fixed soon. Is that issue also exist in the clustered ray `ray start`?
Any workaround for the current situation if I want to limit my actor's RAM / object store memory usage?
What do you think if I make custom resource named `RAM` and I provide my system's RAM value to that custom resource?

**Comment by rkooo567** [comment]
I think when you run `ray start`, it uses just bytes. I am not 100% sure how it works when you specify @ray.remote(memory=[bytes]). @edoakes Do you know if it uses bytes here?

**Comment by muhammadsyahyas** [comment]
Please consider this example:
```python
import ray
from ray import serve
import requests

ray.init()
print(ray.nodes())


@ray.remote(memory=52428800)
def simple_echo():
    return "=============PASSED============="


print(ray.get(simple_echo.remote()))

client = serve.start()


def echo(flask_request):
    return "hello " + flask_request.args.get("name", "serve!")


client.create_backend(
    "hello", echo, ray_actor_options={"memory": 52428800})
client.create_endpoint("hello", backend="hello", route="/hello")

print(client.list_backends())
print(requests.get("http://127.0.0.1:8000/hello").text)
```
Here's the result in my machine:
```
2020-10-07 07:31:32,479	INFO services.py:1166 -- View the Ray dashboard at http://127.0.0.1:8265
[{'NodeID': '8324fb9dab1a397d62947f7a054f488e02111be3', 'Alive': True, 'NodeManagerAddress': '192.168.0.246', 'NodeManagerHostname': 'yahya-Aspire-4750', 'NodeManagerPort': 64217, 'ObjectManagerPort': 36283, 'ObjectStoreSocketName': '/tmp/ray/session_2020-10-07_07-31-31_692952_14819/sockets/plasma_store', 'RayletSocketName': '/tmp/ray/session_2020-10-07_07-31-31_692952_14819/sockets/raylet', 'MetricsExportPort': 47629, 'alive': True, 'Resources': {'accelerator_type:GT': 1.0, 'node:192.168.0.246': 1.0, 'memory': 60.0, 'object_store_memory': 20.0, 'CPU': 8.0, 'GPU': 1.0}}]
=============PASSED=============
(pid=14960) 2020-10-07 07:31:37,473	INFO controller.py:194 -- Starting router with name 'ufmAKJ:SERVE_CONTROLLER_ACTOR:SERVE_PROXY_ACTOR-node:192.168.0.246-0' on node 'node:192.168.0.246-0' listening on '127.0.0.1:8000'
(pid=14956) INFO:     Started server process [14956]
Traceback (most recent call last):
  File "ray_serve_error.py", line 24, in <module>
    "hello", echo, ray_actor_options={"memory": 52428800})
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/api.py", line 29, in check
    return f(self, *args, **kwargs)
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/api.py", line 246, in create_backend
    replica_config))
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/worker.py", line 1428, in get
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(RayServeException): ray::ServeController.create_backend() (pid=14960, ip=192.168.0.246)
  File "python/ray/_raylet.pyx", line 484, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 435, in ray._raylet.execute_task.function_executor
  File "python/ray/_raylet.pyx", line 1404, in ray._raylet.CoreWorker.run_async_func_in_event_loop
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 428, in result
    return self.__get_result()
  File "/usr/lib/python3.7/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 784, in create_backend
    raise e
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 781, in create_backend
    self._scale_replicas(backend_tag, backend_config.num_replicas)
  File "/home/yahya/virtualenvs/tools-3.7/lib/python3.7/site-packages/ray/serve/controller.py", line 554, in _scale_replicas
    num_possible, current_num_replicas + num_possible))
ray.serve.exceptions.RayServeException: Cannot scale backend hello to 1 replicas. Ray Serve tried to add 1 replicas but the resources only allows 0 to be added. To fix this, consider scaling to replica to 0 or add more resources to the cluster. You can check avaiable resources with ray.nodes().
```

That's weird, since I can call my remote function properly when I set the memory resource in `@ray.remote`, but not when set it in `client.create_backend("hello", echo, ray_actor_options={"memory": 52428800})`.

**Comment by muhammadsyahyas** [comment]
@rkooo567 I guess the bug comes from these lines: https://github.com/ray-project/ray/blob/91d0f41a2f7322d11783deb182549dc6e7f20390/python/ray/serve/utils.py#L268-L269

I asked 52428800 **bytes** of memory resource, so it must be `key="memory"` and `count=52428800`, but I guess `node_resource.get(key, 0)` producing value in **chunks**, in my case `60.0`. Since `60.0 - 52428800 < 0`, then my node treated as not feasible, so Serve tried to scale the backend.

**Comment by rkooo567** [comment]
Thanks for the report! cc @simon-mo @edoakes

**Comment by rkooo567** [comment]
I added P2 because I don't know the priority of this task, but it definitely seems to be a bug for me? Please take a look.

**Comment by edoakes** [comment]
@muhammadsyahyas thanks for the report, this is definitely a bug in the resource-checking code because the value returned from `available_resources` is in "chunks" (which is a pretty confusing behavior...). Any chance you'd be willing to contribute a PR to fix this? :) Otherwise we will resolve it soon.

**Comment by RaedShabbir** [comment]
Hello, 

Wanted to confirm if any fixes are coming for this soon, else I can submit a PR to convert from chunks to bytes depending on ray version

**Comment by simon-mo** [comment]
Please feel free to submit a PR!

**Comment by edoakes** [comment]
No longer relevant because we don't use this calculation anymore.
